app/controllers/users_controller.rb

```diff
@@ -866,6 +865,8 @@ def password_reset_update
     # no point doing anything else if we can't even find
     # a user from the token
     if @user
+      raise Discourse::ReadOnly if staff_writes_only_mode? && !@user.staff?
```

Comment:
Why is this line in two places? (session and user)

Reply:
the code in the session controller is used to start the forgot password workflow by entering a username/email and the server will send a token in a forget password email.

the code in the user controller is used to submit a new password.